hazelcast/src/main/java/com/hazelcast/internal/config/override/PropertiesToNodeConverter.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.config.override;
+
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+final class PropertiesToNodeConverter {
+
+    private PropertiesToNodeConverter() {
+    }
+
+    static ConfigNode propsToNode(Map<String, String> properties) {
+        String rootNode = findRootNode(properties);
+
+        ConfigNode root = new ConfigNode(rootNode);
+        for (Map.Entry<String, String> e : properties.entrySet()) {
+            parseEntry(new AbstractMap.SimpleEntry<>(e.getKey().replaceFirst(rootNode + ".", ""), e.getValue()), root);
```

Comment:
No need to create an entry as a parameter, I think it can be:
```
    private static void parseEntry(String key, String value, ConfigNode root) {
        ConfigNode last = root;
        for (String s : key.toLowerCase().split("\\.")) {
            ConfigNode node = last.getChildren().get(s);
            if (node == null) {
                node = new ConfigNode(s, last);
                last.getChildren().put(s, node);
            }
            last = node;
        }

        last.setValue(value);
    }
```


Another alternative is this:
```
static ConfigNode propsToNode(Map<String, String> properties) {
        Map<String, List<BiTuple<String[], String>>> propsByRootNode =
                properties.entrySet().stream()
                          .map(entry -> BiTuple.of(entry.getKey().split("\\."), entry.getValue()))
                          .collect(Collectors.groupingBy(e -> e.element1()[0]));

        if (propsByRootNode.size() > 1) {
            throw new IllegalStateException("parsed config entries have conflicting root node names");
        }
        if (propsByRootNode.size() == 0){
            throw new IllegalStateException("No parsed entries found");
        }

        Entry<String, List<BiTuple<String[], String>>> rootNodeEntry = propsByRootNode.entrySet().iterator().next();

        ConfigNode root = new ConfigNode(rootNodeEntry.getKey());
        for (BiTuple<String[], String> e : rootNodeEntry.getValue()) {
            parseEntry(e.element1, e.element2, root);
        }
        return root;
    }

    private static void parseEntry(String[] keyPath, String value, ConfigNode root) {
        ConfigNode last = root;
        for (int i = 1; i < keyPath.length; i++) {
            ConfigNode node = last.getChildren().get(keyPath[i]);
            if (node == null) {
                node = new ConfigNode(keyPath[i], last);
                last.getChildren().put(keyPath[i], node);
            }
            last = node;
        }

        last.setValue(value);
    }
```

Reply:
With the alternative approach, I've asked myself what can happen if the key is equal to the root name. For instance, this case:
```
HZ_NETWORK_JOIN=whoops
HZ_NETWORK_JOIN_TCPIP_ENABLED=false;
HZ_NETWORK_JOIN_MULTICAST_ENABLED=false
```

Reply:
Any ideas about the above when the key is equal to the root name?